kernel/src/fs/notify/mod.rs

```diff
@@ -0,0 +1,363 @@
+// SPDX-License-Identifier: MPL-2.0
+
+use alloc::{sync::Arc, vec::Vec};
+use core::{
+    any::Any,
+    sync::atomic::{AtomicU32, Ordering},
+};
+
+use bitflags::bitflags;
+use ostd::sync::RwLock;
+
+use crate::{
+    fs::{file_handle::FileLike, path::Path, utils::AccessMode},
+    prelude::*,
+};
+
+use super::utils::{Inode, InodeType};
+
+/// Publishes filesystem events to subscribers.
+///
+/// Each inode has an associated `FsnotifyPublisher` that maintains a list of
+/// subscribers interested in filesystem events. When an event occurs, the publisher
+/// notifies all subscribers whose interested events match the event.
+pub struct FsnotifyPublisher {
+    /// List of fsnotify subscribers.
+    subscribers: RwLock<Vec<Arc<dyn FsnotifySubscriber>>>,
+    /// All interested fsnotify events (aggregated from all subscribers).
+    /// Stored as AtomicU32 for lock-free reads, but accessed as FsnotifyEvents.
+    all_interested_events: AtomicU32,
+}
+
+impl Debug for FsnotifyPublisher {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        let subscribers = self.subscribers.read();
+        write!(
+            f,
+            "FsnotifyPublisher: num_subscribers: {}",
+            subscribers.len()
+        )?;
+        Ok(())
+    }
+}
+
+impl Default for FsnotifyPublisher {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl FsnotifyPublisher {
+    pub fn new() -> Self {
+        Self {
+            subscribers: RwLock::new(Vec::new()),
+            all_interested_events: AtomicU32::new(0),
+        }
+    }
+
+    /// Adds a subscriber to this publisher.
+    pub fn add_subscriber(&self, subscriber: Arc<dyn FsnotifySubscriber>) {
+        let mut subscribers = self.subscribers.write();
+        // Update aggregated events with new subscriber's interested events
+        self.all_interested_events
+            .fetch_or(subscriber.interested_events().bits(), Ordering::Relaxed);
+        subscribers.push(subscriber);
+    }
+
+    /// Removes a subscriber from this publisher.
+    pub fn remove_subscriber(&self, subscriber: &Arc<dyn FsnotifySubscriber>) {
+        let mut subscribers = self.subscribers.write();
+        subscribers.retain(|m| !Arc::ptr_eq(m, subscriber));
+        subscriber.deliver_event(FsnotifyEvents::IN_IGNORED, None);
+
+        // Recalculate aggregated events from remaining subscribers
+        self.recalc_interested_events(&subscribers);
```

Comment:
To be correct in grammar, change all "interested events" to "interesting events".